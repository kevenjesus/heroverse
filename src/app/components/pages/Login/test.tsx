import { act, fireEvent, render, waitFor } from "@testing-library/react"
import LoginPage from "."
import '@testing-library/jest-dom'

const pushMock = jest.fn()

jest.mock('next/navigation', () => ({
    useRouter: () => ({
        push: pushMock
    })
}))


describe("LoginPage", () => {
    afterEach(() => {
        sessionStorage.removeItem("isLoggin")
    })
    it("should render", () => {
        const { container } = render(<LoginPage />)
        expect(container).toMatchSnapshot()
    })

    it("should redirect to home page when user is already logged in", () => {
        sessionStorage.setItem("isLoggin", "true")
        const { container } = render(<LoginPage />)
        expect(container.firstChild).toBe(null)
    })

    it("should display a message erro with field login is empty", () => {
        const { getByRole, getByText } = render(<LoginPage />)
        const button = getByRole('button', {name: 'Entrar'})
        fireEvent.click(button)
        expect(getByText('Por favor. Informe o campo login')).toBeInTheDocument()
    })

    it("should display a message erro with field password is empty", () => {
        const { getByPlaceholderText, getByRole, getByText } = render(<LoginPage />)
        const login = getByPlaceholderText("Login")
        fireEvent.change(login, {target: {value: 'user'}})
        const button = getByRole('button', {name: 'Entrar'})
        fireEvent.click(button)
        expect(getByText('Por favor. Informe o campo senha')).toBeInTheDocument()
    })

    it("should display a message erro with login is invalid", () => {
        const { getByPlaceholderText, getByRole, getByText } = render(<LoginPage />)
        const login = getByPlaceholderText("Login")
        const password = getByPlaceholderText("Senha")

        fireEvent.change(login, {target: {value: 'user'}})
        fireEvent.change(password, {target: {value: 'pass'}})

        const button = getByRole('button', {name: 'Entrar'})
        fireEvent.click(button)
        expect(getByText('Usuário e/ou senha inválidos')).toBeInTheDocument()
    })

    it("should display a message success with login is valid", () => {
        const { getByPlaceholderText, getByRole, getByText } = render(<LoginPage />)
        const login = getByPlaceholderText("Login")
        const password = getByPlaceholderText("Senha")

        fireEvent.change(login, {target: {value: 'globalthings'}})
        fireEvent.change(password, {target: {value: 'globalthings@123'}})

        const button = getByRole('button', {name: 'Entrar'})
        fireEvent.click(button)
        waitFor(() => {
            expect(getByText('Bem-vindo(a), voce será redirecionado(a)')).toBeInTheDocument()
        })
    })

    it('should redirect to home page after successful login', () => {      
        const { getByPlaceholderText, getByText } = render(<LoginPage />);

        process.env.NEXT_PUBLIC_LOGIN="globalthings"
        process.env.NEXT_PUBLIC_PASSWORD="globalthings@123"
        
        const loginInput = getByPlaceholderText('Login');
        const passwordInput = getByPlaceholderText('Senha');
        const submitButton = getByText('Entrar');
        
        fireEvent.change(loginInput, { target: { value: 'globalthings' } });
        fireEvent.change(passwordInput, { target: { value: 'globalthings@123' } });
        
        jest.useFakeTimers();
        fireEvent.click(submitButton);

        jest.runAllTimers();
        expect(pushMock).toBeCalled()
        jest.useRealTimers();

      });
      
})