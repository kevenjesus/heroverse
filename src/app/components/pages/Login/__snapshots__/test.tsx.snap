// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage should render 1`] = `
<div>
  <div
    class="sc-hLseeU gBExbG"
  >
    <img
      alt="Hero Verse"
      class="sc-eDDNvR gfOpcO"
      src="/heroverse.png"
      width="160"
    />
    <form
      class="sc-jTrPJq gfIiLF"
    >
      <input
        class="sc-dmqHEX fXVpfR"
        name="login"
        placeholder="Login"
        role="textbox"
        type="text"
        value=""
      />
      <input
        class="sc-dmqHEX fXVpfR"
        name="password"
        placeholder="Senha"
        role="textbox"
        type="password"
        value=""
      />
      <button
        class="sc-gLDzan iEQYAJ"
        type="submit"
      >
        <span>
          Entrar
        </span>
      </button>
    </form>
    <div
      style="padding: 50px 0px;"
    >
      <p
        class="sc-iAEyYk jnoSnn"
      >
        Hero Verse - Todos os direitos reservados
      </p>
    </div>
  </div>
</div>
`;
